### PR TITLE
Fix _gcdcache

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by jksuom** [comment]
+1

**Comment by asmeurer** [comment]
Can you explain what this fix does? The code seems identical to me.

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def igcd(*args):
             a = _gcdcache[(a, b)]
         except KeyError:
             a, b = as_int(a), as_int(b)
-
+            t = (a, b)
```
Comment (isuruf) [inline sympy/core/numbers.py]: `a, b` are modified in the lines below and when code reaches `_gcdcache`, `b` is 0 and therefore cache only `(a, 0) -> a` which is a trivial result.

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -161,7 +161,7 @@ def igcd(*args):
             a = _gcdcache[(a, b)]
         except KeyError:
             a, b = as_int(a), as_int(b)
```
Comment (asmeurer) [inline sympy/core/numbers.py]: If anything, write `t = as_int(a), as_int(b)` here. But, again, I don't see why this change is necessary. It just seems to create an unnecessary variable. Is creating the tuple slow (but the way you've got it, you still create the tuple twice)?

File: sympy/core/numbers.py (hunk 1 above)
Comment (asmeurer) [inline sympy/core/numbers.py]: Ah, I see now. Can you please indicate this in the commit message?

Comment (isuruf) [inline sympy/core/numbers.py]: Done

**Comment by smichr** [comment]
I have some changes in the pipe, too, at https://github.com/smichr/sympy/blob/fractions/sympy/core/numbers.py

**Comment by smichr** [comment]
The thing that I noticed (that I didn't think was too much of an issue) was that the un-as_int'ified values were being used as keys for the caching. Is this what you noticed, too?

**Comment by asmeurer** [comment]
That's a good point. It should just `as_int` the args at the top of the function.

**Comment by isuruf** [comment]
@smichr, let me know when you are ready with your branch for me to review.
Btw, `setdefault` is not a lazy evaluation, so you would need to change it a bit.

**Comment by smichr** [comment]
I'll update tomorrow morning. I see what you mean about the similarity of igcd and gcd.

**Comment by smichr** [comment]
To benchmark any changes I have been using

```
from sympy.core.number import _gcdcache as G
from time import time
def go(n):
 for a in range(n):
    for b in range(n):
        g=igcd(a,b)

G.clear();t=time();go(input());time()-t
```

But perhaps that is not a good test -- should `n` random pairs be tested instead?

Almost any change I make is worse that what `master` has. I don't know a lot about smart caching. I know that you want to get likely frequent or expensive entries cached but I also know there is a cost to even figuring out if such an entry should be cached...and caching everything leads to a bloated cache with longer lookups (or maybe that's not an issue?) and memory usage.
